File: 03-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if (len(s) == 0): return 0
        
        contained = set()
        result = 0

        for element in s:
            if (element in contained): break;

            result += 1;
            contained.add(element)

        return max(self.lengthOfLongestSubstring(s[1:]), result)

def main():
    sol =  Solution();
    print(sol.lengthOfLongestSubstring("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~ abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"))

if (__name__ == "__main__"):
    main()

File: 03-longest-substring-without-repeating-characters.cpp
#include <iostream>
#include <unordered_map>

class Solution {
 public:
  int lengthOfLongestSubstring(std::string s) {
    if (s.length() == 0) {
      return 0;
    }

    std::unordered_map<char, int> map;
    int max = 0;

    for (int i = 0, j = 0; i < s.length(); ++i) {
      if (map.find(s[i]) != map.end()) {
        j = std::max(j, map.find(s[i])->second + 1);
      }
      map.insert_or_assign(s[i], i);
      max = std::max(max, i - j + 1);
    }

    return max;
  }
};

int main() {
  Solution solution;
  std::cout << solution.lengthOfLongestSubstring("abcabcbb") << std::endl;
}
